### PR TITLE
feat(protocol): add aggregated sgx verify test

### PR DESCRIPTION
**Pull request by smtmfft**

### CODE REVIEW

File: packages/protocol/contracts/layer1/verifiers/SgxVerifier.sol [hunk 1]
```diff
@@ -172,14 +172,49 @@ contract SgxVerifier is EssentialContract, IVerifier {
...
+        // 4 bytes + 20 bytes + 20 bytes + 65 bytes (signature) = 109
+        if (_proof.data.length != 109) revert SGX_INVALID_PROOF();
+
+        uint32 id = uint32(bytes4(_proof.data[:4]));
+        address oldInstance = address(bytes20(_proof.data[4:24]));
+        address newInstance = address(bytes20(_proof.data[24:44]));
+        bytes memory signature = _proof.data[44:];
+
+        // Collect public inputs
+        bytes32[] memory public_inputs = new bytes32[](_ctxs.length + 2);
```
Comment (dantaik) [inline packages/protocol/contracts/layer1/verifiers/SgxVerifier.sol:191]:
```suggestion
        bytes32[] memory publicInputs = new bytes32[](_ctxs.length + 2);
```

File: packages/protocol/contracts/layer1/verifiers/SgxVerifier.sol [hunk 2]
```diff
@@ -172,14 +172,49 @@ contract SgxVerifier is EssentialContract, IVerifier {
...
+        // First public input is the current instance public key
+        public_inputs[0] = bytes32(uint256(uint160(oldInstance)));
+        public_inputs[1] = bytes32(uint256(uint160(newInstance)));
+        // All other inputs are the block program public inputs (a single 32 byte value)
+        for (uint256 i = 0; i < _ctxs.length; i++) {
```
Comment (dantaik) [inline packages/protocol/contracts/layer1/verifiers/SgxVerifier.sol:196]:
```suggestion
        for (uint256 i = 0; i < _ctxs.length; ++i) {
```

File: packages/protocol/contracts/layer1/verifiers/SgxVerifier.sol [hunk 3]
```diff
@@ -172,14 +172,49 @@ contract SgxVerifier is EssentialContract, IVerifier {
...
+            // TODO: For now this assumes the new instance public key to remain the same
+            TaikoData.Transition memory tran = _ctxs[i].tran;
+            address prover = _ctxs[i].prover;
+            bytes32 metaHash = _ctxs[i].metaHash;
+            public_inputs[i + 2] = LibPublicInput.hashPublicInputs(
+                tran, address(this), newInstance, prover, metaHash, taikoChainId()
```
Comment (dantaik) [inline packages/protocol/contracts/layer1/verifiers/SgxVerifier.sol:202]: use _ctxs[i].metaHash and _ctxs[i].prover directly here?

**Review by dantaik — APPROVED** [review]

File: packages/protocol/contracts/layer1/verifiers/Risc0Verifier.sol [hunk 4]
```diff
@@ -80,13 +83,50 @@ contract Risc0Verifier is EssentialContract, IVerifier {
...
+        if (!isImageTrusted[blockImageId]) {
+            revert RISC_ZERO_INVALID_IMAGE_ID();
+        }
+
+        // Collect public inputs
+        bytes32[] memory public_inputs = new bytes32[](_ctxs.length + 1);
+        // First public input is the block proving program key
+        public_inputs[0] = blockImageId;
+        // All other inputs are the block program public inputs (a single 32 byte value)
+        for (uint256 i = 0; i < _ctxs.length; i++) {
```
Comment (dantaik) [inline packages/protocol/contracts/layer1/verifiers/Risc0Verifier.sol:109]:
```suggestion
        for (uint256 i; i < _ctxs.length; ++i) {
```

File: packages/protocol/contracts/layer1/verifiers/Risc0Verifier.sol [hunk 5]
```diff
@@ -80,13 +83,50 @@ contract Risc0Verifier is EssentialContract, IVerifier {
...
+        (bytes memory seal, bytes32 blockImageId, bytes32 aggregationImageId) =
+            abi.decode(_proof.data, (bytes, bytes32, bytes32));
+
+        // Check if the aggregation program is trusted
+        if (!isImageTrusted[aggregationImageId]) {
+            revert RISC_ZERO_INVALID_IMAGE_ID();
+        }
+        // Check if the block proving program is trusted
+        if (!isImageTrusted[blockImageId]) {
+            revert RISC_ZERO_INVALID_IMAGE_ID();
```
Comment (dantaik) [inline packages/protocol/contracts/layer1/verifiers/Risc0Verifier.sol:101]: use two different error codes?

**Review by dantaik — CHANGES_REQUESTED** [review]

**Review by dantaik — APPROVED** [review]

**Review by YoGhurt111 — APPROVED** [review]

**Review by YoGhurt111 — APPROVED** [review]

**Review by davidtaikocha — APPROVED** [review]

**Pull request merged**